### DISCUSSION TITLE
Best practices for ArgoCD application controller resources

### DISCUSSION THREAD

**Discussion by musabmasood** [argoproj/argo-cd#13595] [Q&A]
We have a single instance ArgoCD managing 4 clusters (1 of them is prod).

The CPU utilization so different on them ( as expected per shard ).

![image](https://github.com/argoproj/argo-cd/assets/8170066/04498093-f5da-43d1-b8b9-c906ec285bd7)

![image](https://github.com/argoproj/argo-cd/assets/8170066/5ef612c5-688f-4d12-8c0a-d7a57f28d827)

![image](https://github.com/argoproj/argo-cd/assets/8170066/54bf46d7-efe0-440f-9fab-e05c7903df44)

![image](https://github.com/argoproj/argo-cd/assets/8170066/24eebd53-b8df-48e2-833e-efc0f61d6880)

We tried scaling them up to 6 replicas (hence the numbering) but its useless. 

Any other best practices out there other than setting the CPU to maybe p90 of the highest load shard?

**Comment by jaideepr97** [comment]
Hi @musabmasood 

Based on my understanding, the current mechanism for distributing clusters across shards is not very sophisticated or predictable (and doesn't guarantee equitable distribution of load between shards)

That may be why you see large swings between shard activity and also why scaling up didn't seem to help (possibly because the new shards did not get assigned any clusters) 

That being said, there is active ongoing work to introduce better load distribution algorithms in the application controller 
(like https://github.com/argoproj/argo-cd/pull/13221 and https://github.com/argoproj/argo-cd/pull/13018)
